extensions/amp-mega-menu/0.1/amp-mega-menu.js

```diff
@@ -0,0 +1,363 @@
+import {
+  closestAncestorElementBySelector,
+  isRTL,
+  scopedQuerySelector,
+  scopedQuerySelectorAll,
+  tryFocus,
+} from '../../../src/dom';
+import {dev, userAssert} from '../../../src/log';
+import {isExperimentOn} from '../../../src/experiments';
+import {mod} from '../../../src/utils/math';
+import {setModalAsClosed, setModalAsOpen} from '../../../src/modal';
+import {toArray} from '../../../src/types';
+
+/** @const {string} */
+const TAG = 'amp-mega-menu';
+
+/** @const {string} */
+const ARIA_LABEL_CLOSE = 'Close the menu';
+
+/**
+ * A mega menu component suitable for displaying large collections of links on
+ * desktop/tablets. An amp-list can be nested inside this component to enable
+ * template rendering via dynamically fetched data.
+ */
+export class AmpMegaMenu extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Array<!Element>} */
+    this.items_ = null;
+
+    /** @private {number} */
+    this.itemCount_ = 0;
+
+    /** @private {?Element} */
+    this.expandedItem_ = null;
+
+    /** @private {?Element} */
+    this.maskElement_ = null;
+
+    /** @private @const {!Document} */
+    this.document_ = this.win.document;
+
+    /** @private @const {!Element} */
+    this.documentElement_ = this.document_.documentElement;
+
+    /** @private {?../../../src/service/action-impl.ActionService} */
+    this.action_ = null;
+
+    /** @private {number|string} */
+    this.prefix_ = element.id || Math.floor(Math.random() * 100);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout === Layout.FIXED_HEIGHT;
+  }
+
+  /** @override */
+  buildCallback() {
+    // TODO(#24898): remove this assert when cleaning up experiment post launch.
+    userAssert(
+      isExperimentOn(this.win, 'amp-mega-menu'),
+      'Turning on the amp-mega-menu experiment is necessary to use the amp-mega-menu component.'
+    );
+
+    this.action_ = Services.actionServiceForDoc(this.element);
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.registerMenuItems_();
+    // items may not be present after build if dynamically rendered via amp-list,
+    // in which case register them after DOM update instead.
+    this.element.addEventListener(AmpEvents.DOM_UPDATE, () => {
+      this.registerMenuItems_();
+    });
+
+    const mask = this.document_.createElement('div');
+    mask.classList.add('i-amphtml-mega-menu-mask');
+    mask.setAttribute('aria-hidden', 'true');
+    // append mask to header so that all children of header appear above the mask
+    const maskParent =
+      closestAncestorElementBySelector(this.element, 'header') || this.element;
+    maskParent.classList.add('i-amphtml-mega-menu-mask-parent');
+    maskParent.appendChild(mask);
+    this.maskElement_ = mask;
+
+    this.documentElement_.addEventListener('click', () => this.collapse_());
+    this.documentElement_.addEventListener('keydown', event => {
```

Comment:
Do these events get cleaned up anywhere besides page deconstruction?

Reply:
Good catch! We should unregister them in `unlayoutCallback`.

Reply:
filed #25047 to make the change and verify via unit test.